Mathlib/Combinatorics/DyckWord.lean

```diff
@@ -0,0 +1,424 @@
+/-
+Copyright (c) 2024 Jeremy Tan. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jeremy Tan
+-/
+import Mathlib.Combinatorics.Catalan
+import Mathlib.Data.List.Indexes
+
+/-!
+# Dyck words
+
+## Main definitions
+
+* `DyckPath`: a list of `Step`s, each of which may only be `U`p or `D`own.
+* `IsBalanced`: the property defining Dyck words. Translating `U` and `D` to `(` and `)`,
+  the resulting brackets are balanced.
+
+## Main results
+
+* `treeEquiv`: equivalence between Dyck words and rooted binary trees.
+* `finiteTreeEquiv`: equivalence between Dyck words of length `2 * n` and rooted binary trees
+  with `n` internal nodes.
+* `treesOfNodesEq_card_eq_catalan`: there are `catalan n` Dyck words of length `2 * n`.
+-/
+
+
+/-- A `Step` is either `U` or `D`, corresponding to `(` and `)` respectively. -/
+inductive Step
+  | U : Step
+  | D : Step
+  deriving Inhabited, DecidableEq
+
+/-- A `DyckPath` is a `List` of `Step`s. -/
+def DyckPath := List Step deriving Append
+
+namespace DyckPath
+
+open Step List
+
+variable {p : DyckPath}
+
+lemma count_U_add_count_D_eq_length : p.count U + p.count D = p.length := by
+  symm; convert p.length_eq_countP_add_countP (· == U)
+  rw [count]; congr!; rename_i s
+  cases' s <;> simp
+
+variable (p)
+
+/-- A `DyckPath` is balanced if its running `U`-count always exceeds its running `D`-count
+and both counts are equal at the end. -/
+def IsBalanced (p : DyckPath) : Prop :=
+  p.count U = p.count D ∧ ∀ i, (p.take i).count D ≤ (p.take i).count U
+
+section Suf
+
+variable (h : p.IsBalanced) (hl : p ≠ [])
+
+lemma step_cases (s : Step) : s = U ∨ s = D := by cases' s <;> tauto
```

Comment:
Can this lemma be autogenerated with `@[mk_iff]` on `Step`?

Reply:
`mk_iff only applies to prop-valued declarations`